server/base/src/main/java/org/apache/accumulo/server/zookeeper/TransactionWatcher.java

```diff
@@ -128,44 +131,59 @@ public boolean transactionComplete(String type, long tid) throws Exception {
    * that task is not run and a debug message is logged indicating the task was ignored.
    */
   public void runQuietly(String ztxBulk, long tid, Runnable task) {
-    synchronized (counts) {
-      try {
-        if (!arbitrator.transactionAlive(ztxBulk, tid)) {
-          log.debug("Transaction " + tid + " of type " + ztxBulk + " is no longer active.");
-          return;
-        }
-      } catch (Exception e) {
-        log.warn("Unable to check if transaction " + tid + " of type " + ztxBulk + " is alive ", e);
+
+    Lock l = stripedLocks.get(tid);
+    l.lock();
+    try {
+      if (!arbitrator.transactionAlive(ztxBulk, tid)) {
+        log.debug("Transaction " + tid + " of type " + ztxBulk + " is no longer active.");
         return;
       }
-      increment(tid);
+    } catch (Exception e) {
+      log.warn("Unable to check if transaction " + tid + " of type " + ztxBulk + " is alive ", e);
+      return;
     }
+    increment(tid);
     try {
       task.run();
     } finally {
       decrement(tid);
+      l.lock();
     }
+
```

Comment:
The locking in this method doesn't look right. There are two locks, no unlocks, and the locks are never released in a finally block.

Reply:
Please check the latest commit. I sent the wrong version.

Reply:
Latest looks better. Thanks.